### ISSUE TITLE
detect "offline", warn, and build package plan from local cache

### ISSUE THREAD

**Issue by ijstokes** [conda/conda#996]
tl;dr: avoid borked environments when pkg plan calls for pkg download (but offline, so download is impossible) by trying to create package plan just from local package cache.

When creating an environment while "offline" (e.g. under circumstances where some or all pkg repositories are inaccessible) the user should be warned and given the option to proceed just with locally cached packages in the package plan.  Right now under these circumstances environment creation just fails if the package plan calls for some packages to be downloaded, leaving a borked env directory.

Commentary:

I have a stack (hundreds) of conda packages in my local package cache.  I'm offline and not about to be online for some time.  I want to create a new environment.  Somehow conda has enough offline information to identify that for some dependencies there are newer packages available to satisfy my package requests, but they will need to be downloaded.  I can see this probably isn't going to work, but select "Yes" to proceed with the package plan (what other option do I have?)

The appropriate environment directory is created, but then conda immediately chokes because it can't download the newer dependency packages it has identified as being required for the package plan.  I think "maybe --use-index-cache will solve this", but also no luck.

I'm able to dig around my package cache and see which are the latest versions I have locally of these dependency packages, then explicitly specify these (earlier) versions and hope for the best.  This works.  I get exactly the same versions of the packages I'm actually interested in.  This makes me think there could (and perhaps "should") be a way for conda to identify that it is "offline" and then build a package plan only off the locally available packages.

This seems like it would be a common use case in situations where all packages come from a locally approved package source, or in situations where many anaconda users are often in an "isolated" or "offline" mode.

**Comment by ijstokes** [comment]
I just came back to this bug.  I think there is a smaller issue here which could more accurately be described as a bug: the package plan that is created *cannot be enacted* (offline), but in attempting to act on it conda will choke and die (stacktrace), leaving a bogus half-created conda env behind.  This is clearly undesirable behavior.  If the repo is unreachable *at the start* then there should be no attempt to enact a package plan and create an env which includes package downloads that will never happen.

Or put differently, perhaps the sequence needs to be reworked in conda: create the package plan, download any packages necessary, *then* try to create the environment.  Right now it appears that the environment is created first, so if there is a failure during the download process (and for sure there are *lots* of reasons why this might happen) then a duff env is going to be left behind.

**Comment by ijstokes** [comment]
@groutr is my last comment (just above) solved in latest conda releases?

@mcg1969 has your re-working of conda and perhaps "clone" addressed the other issue I raise about package plans, and the behavior when "offline" if the plan has identified the need to download?

 @kalefranz do you have a perspective on this issue or its importance/priority?

**Comment by kalefranz** [comment]
There is now an `--offline` flag, or more permanently set `conda config --set offline true`.  It completely disable conda from making network requests.  So what's available to conda is any channel that's at a `file://` url, and what's in the package cache.
